### PR TITLE
oSPARC viewers

### PR DESCRIPTION
**Pull request by ignapas**
# Description

Adds the possibility for the users to open files in oSPARC.
* New option in file actions dropdown in the Files tab (dataset detail)
* Option opens a new modal window to select the viewer
Peek 2021-03-03 12-24
* First option selected by default
* Link to more information about oSPARC

## Type of change

New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

There are no automated tests for this feature. Manual testing has been done.
## How to test
* Needs https://github.com/nih-sparc/sparc-api/pull/34 running
* Go to a dataset detail page and open the files tab
* Go to one file and open the options (button on the right)
* If the file is viewable in oSPARC, the corresponding option will appear
* Clicking on the option opens a modal window to select the file
* A new window opens and the file is displayed within the oSPARC platform
* Repeat the operation for some other files
* You can check the available file viewers in https://osparc.io/v0/viewers

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: components/FilesTable/FilesTable.vue [hunk 1]
```diff
@@ -170,10 +185,13 @@ import {
 } from 'ramda'
 
 import BfDownloadFile from '@/components/BfDownloadFile/BfDownloadFile'
+import OsparcFileViewersDialog from '@/components/FilesTable/OsparcFileViewersDialog.vue'
 
 import FormatStorage from '@/mixins/bf-storage-metrics/index'
 import RequestDownloadFile from '@/mixins/request-download-file'
 
+import { extractExtension } from '~/pages/data/utils'
```
Comment (cameronbaney) [inline components/FilesTable/FilesTable.vue:193]:
```suggestion
import { extractExtension } from '@/pages/data/utils'
```

File: components/FilesTable/FilesTable.vue [hunk 2]
```diff
@@ -151,6 +160,12 @@
           </template>
         </el-table-column>
       </el-table>
+      <osparc-file-viewers-dialog
+        :open="dialogSelectedFile !== null"
+        :onClose="() => setDialogSelectedFile(null)"
```
Comment (cameronbaney) [inline components/FilesTable/FilesTable.vue:165]: This works, but the Vue way is to pass props down and [emit events up](https://vuejs.org/v2/guide/components-custom-events.html). Not saying this needs to change.

More info: https://medium.com/js-dojo/passing-functions-as-props-an-anti-pattern-in-vue-js-b542fc0cf5d

File: components/FilesTable/OsparcFileViewersDialog.vue [hunk 3]
```diff
@@ -0,0 +1,105 @@
+<template>
```
Comment (cameronbaney) [inline components/FilesTable/OsparcFileViewersDialog.vue:1]: Formatting is off, which should have been caught by the linter. I'm making a note to check into the linting checks. Checking `yarn lint` shows quite a few linting errors in the code base, so something is not right.

File: components/DatasetDetails/DatasetFilesInfo.vue [hunk 4]
```diff
@@ -30,6 +30,9 @@ export default {
     datasetDetails: {
       type: Object,
       default: () => {}
+    },
+    osparcViewers: {
+      type: Object
```
Comment (cameronbaney) [inline components/DatasetDetails/DatasetFilesInfo.vue:35]:
```suggestion
      type: Object,
      default: () => {}
```

File: components/FilesTable/OsparcFileViewersDialog.vue [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+export default {
+  name: "OsparcFileViewersDialog",
+  components: { BfButton, BfDialogHeader, DialogBody },
+  props: {
+    selectedFile: {
+      type: Object
+    },
+    viewers: {
+      type: Object
+    },
```
Comment (cameronbaney) [inline components/FilesTable/OsparcFileViewersDialog.vue:43]:
```suggestion
    selectedFile: {
      type: Object,
      default: () => {}
    },
    viewers: {
      type: Object,
      default: () => {}
    },
```

File: pages/data/utils.ts [hunk 6]
```diff
@@ -135,3 +135,19 @@ export const handleSortChange = (
 
   fetchResults()
 }
+
+/**
+ * Gets the file extension for a file or path
+ * @param path Path or filename
+ * @returns String containing the file extension
+ */
+export const extractExtension = (
```
Comment (cameronbaney) [inline pages/data/utils.ts:144]: Nice

**Review by cameronbaney — COMMENTED** [review]
This is looking great. I left a few minor comments, and can't seem to get any data from the sparc-api `/get_osparc_data` endpoints. Even https://osparc.io/v0/viewers is showing no data. This means I can't fully test, but everything on the sparc-app side of things is looking good.

File: components/FilesTable/FilesTable.vue (hunk 2 above)
Comment (ignapas) [reply components/FilesTable/FilesTable.vue]: I made the change, please check that is the way to go!

Comment (cameronbaney) [reply components/FilesTable/FilesTable.vue]: @ignapas Sorry it took so long to review but this is looking great now!

**Review by cameronbaney — APPROVED** [review]

**Pull request merged**